File: frontend/src/types/pipeline_run.ts
export interface PipelineRun {
  run_id: number;
  started_at: string;
  finished_at: string | null;
  status: "running" | "completed" | "failed";
  input_files: string[] | null;
  rows_processed: number;
  rows_rejected: number;
  rejection_rate: number;
  error_message: string | null;
}

File: api/schemas/pipeline_run.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional
import json

class PipelineRunResponse(BaseModel):
    run_id: int
    started_at: datetime
    finished_at: Optional[datetime] = None
    status: str
    input_files: Optional[list[str]] = None
    rows_processed: int
    rows_rejected: int
    rejection_rate: float
    error_message: Optional[str] = None

    @classmethod
    def from_row(cls, row: dict) -> "PipelineRunResponse":
        data = dict(row)
        if data.get("input_files"):
            try:
                data["input_files"] = json.loads(data["input_files"])
            except:
                data["input_files"] = None
        return cls(**data)
